src/profile/helpers/editData.ts

```diff
@@ -0,0 +1,202 @@
+
+export type EditData = {
+  readonly id: string;
+  readonly value: EditDataValue;
+  readonly primary?: boolean;
+  readonly saving: SaveType;
+};
+
+export type EditFunctions = {
+  create: (newProfileData: EditDataProfileSource) => EditData;
+  getEditData: () => EditData[];
+};
+
+function isMultiItemDataType(dataType: EditDataType): boolean {
+  return !(
+    dataType === basicDataType || dataType === additionalInformationType
+  );
+}
+
+function getValueProps(dataType: EditDataType): string[] {
+  if (dataType === 'phones') {
+    return ['phone'];
+  } else if (dataType === 'emails') {
+    return ['email'];
+  } else if (dataType === 'addresses') {
+    return ['postalCode', 'address', 'city', 'countryCode'];
+  } else if (dataType === basicDataType) {
+    return ['firstName', 'nickname', 'lastName'];
+  } else {
+    return ['language'];
+  }
+}
+
+function getPrimaryProp(
+  dataType: EditDataType
+): 'primaryPhone' | 'primaryEmail' | 'primaryAddress' {
+  if (dataType === 'phones') {
+    return 'primaryPhone';
+  } else if (dataType === 'emails') {
+    return 'primaryEmail';
+  } else {
+    return 'primaryAddress';
+  }
+}
+
+function getPrimary(
+  profileData: ProfileData,
+  dataType: EditDataType
+): Partial<MultiItemProfileNode> | null {
+  const primaryProp = getPrimaryProp(dataType);
+  return profileData[primaryProp];
+}
+
+function createNewProfileNode<T extends MultiItemProfileNode>(
+  dataType: EditDataType,
+  overrides?: Partial<T>
+): T {
+  return {
+    ...(formConstants.EMPTY_VALUES[dataType] as T),
+    ...overrides,
+  };
+}
+
+function pickValue(
+  profileDataItem: EditDataProfileSource,
+  dataType: EditDataType
+): EditDataValue {
+  const pickProps = getValueProps(dataType);
+  return _.pick(profileDataItem, pickProps) as EditDataValue;
+}
+
+function create(
+  profileData: EditDataProfileSource,
+  dataType: EditDataType,
+  overrides?: { value?: EditDataValue; saving?: SaveType }
+): EditData {
+  return {
+    id: profileData.id,
+    primary: !!(profileData as MultiItemProfileNode).primary,
+    value: pickValue(profileData, dataType),
+    saving: undefined,
+    ...overrides,
+  };
+}
+
+function pickSources(
+  profileData: ProfileData,
+  dataType: EditDataType
+): EditDataProfileSource[] {
+  if (!isMultiItemDataType(dataType)) {
+    const values = pickValue(profileData, dataType) as BasicDataValue &
+      AdditionalInformationValue;
+    return [{ ...values, id: profileData.id }];
+  } else {
+    const getter =
+      dataType === 'phones'
+        ? getPhonesFromNode
+        : dataType === 'emails'
+        ? getEmailsFromNode
+        : getAddressesFromNode;
+    const nodes: MultiItemProfileNode[] = getter({ myProfile: profileData });
+    const primaryData = getPrimary(profileData, dataType);
+    if (primaryData) {
+      const primaryAsNode = createNewProfileNode<MultiItemProfileNode>(
+        dataType,
+        primaryData
+      );
+      nodes.unshift(primaryAsNode);
+    }
```

Comment:
😮 Has the myProfileQuery always returned primary nodes in the node list too? The old Profile UI code seems odd in that case. And the old query result used in tests is incorrect. Primary nodes are not in the associated node lists. (Fixed now). 

Profiili API seems to return primary nodes correctly as #0 in node lists. In old code, updateMutationVariables uses primaries, but otherwise they are not needed in Profile UI. ..if they are always the first node in the node list. Is this guaranteed? The query does not seem to support sorting, so UI cannot set it. 

I'd prefer deleting primaryAddress, primaryEmail and primaryPhone in another PR.

I added a new param to all get*FromNode, so EditData doesn't need to play with primaries.

Reply:
I don't know if the primaries have _always_ been included in the lists. But they have been there at least as long as I have been around.

Currently it's guaranteed in API code that the primary comes first in the list, so it doesn't happen by accident. But I'm not sure if it's a public API guarantee (meaning, there would be tests ensuring it remains so, and the behaviour would be documented). We think we can make it a hard guarantee, documented and all.